pkg/webhooks/errors_test.go

```diff
@@ -166,6 +166,41 @@ func TestIsAdmissionInvalidRemovingStorage(t *testing.T) {
 	}
 }
 
+func TestIsAdmissionInvalidExternalListenerPort(t *testing.T) {
+	testCases := []struct {
+		testName  string
+		fieldErrs field.ErrorList
+		want      bool
+	}{
+		{
+			testName: "field.Invalid_externalListeners.[0]",
+			fieldErrs: append(field.ErrorList{}, field.Invalid(field.NewPath("spec").Child("listenersConfig").Child("externalListeners").Index(0).Child("externalStartingPort"), int32(79090),
+				invalidExternalListenerStartingPortErrMsg+": "+fmt.Sprintf("ExternalListener '%s' would generate invalid port numbers (not between 1 and 65535) for brokers %v", "test-external1", []int32{0, 1, 2}))),
+			want: true,
+		},
+		{
+			testName: "field.Invalid_externalListeners.[1]_wrong-error-message",
+			fieldErrs: append(field.ErrorList{}, field.Invalid(field.NewPath("spec").Child("listenersConfig").Child("externalListeners").Index(1).Child("externalStartingPort"), int32(59090),
+				"wrong-error-message"+": "+fmt.Sprintf("ExternalListener '%s' would generate invalid port numbers (not between 1 and 65535) for brokers %v", "test-external1", []int32{901, 902}))),
+			want: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.testName, func(t *testing.T) {
+			kafkaCluster := banzaicloudv1beta1.KafkaCluster{ObjectMeta: metav1.ObjectMeta{Name: "test-KafkaCluster"}}
+
+			err := apierrors.NewInvalid(
+				kafkaCluster.GetObjectKind().GroupVersionKind().GroupKind(),
+				kafkaCluster.Name, tc.fieldErrs)
+
+			if got := IsAdmissionInvalidExternalListenerPort(err); got != tc.want {
+				t.Errorf("Check External Listener Port Error message. Expected: %t ; Got: %t", tc.want, got)
```

Comment:
Something that I learned from Patrik: https://github.com/banzaicloud/koperator/pull/928#discussion_r1103573247

The same comment goes to other `t.Errorf` in this PR

Reply:
Like 4f1fcdc1eaa51f5764a69e39c3ad860e667b8a55 ?

One more thing:   
I chose to do matching with the `ElementsMatch` function because order of the `*field.Error` elements in the `field.ErrorList` doesn't matter (IMO) even though it is deterministic. With the current test cases the `Equal` function could be used just as well.  
Do we prefer `Equal` instead ? (to be 100% precise)

Reply:
> Do we prefer Equal instead ? (to be 100% precise)

Compared to `t.Errorf`, yes, since it's a bit cleaner

Reply:
I was unclear.  
I meant : do we prefer `require.Equal()` or `require.ElementsMatch()` ?

Reply:
> do we prefer `require.Equal()` or `require.ElementsMatch()` ?

To make a good decision on this, I checked what these two functions for, and found that
- `require.Equal()` is to "assert that two objects are equal", and "Pointer variable equality is determined based on the equality of the referenced values (as opposed to the memory addresses). Function equality cannot be determined and will always fail."
- And `require.ElementsMatch()` is to "assert that the specified listA(array, slice...) is equal to specified listB(array, slice...) ignoring the order of the elements. If there are duplicate elements, the number of appearances of each of them in both lists should match."

So, `require.ElementsMatch()` is a better choice if you are comparing two slices and arrays while ignoring the order of the elements. And `require.Equal()` is to compare two objects in general

But above is just my understanding after diving into a bit of the source code of those two methods, it'd be great if @pregnor can correct me if I am mistaken somewhere

Reply:
In this line in the `errors_test.go` file `require.Equal()` is the way to go.  

The question about `require.Equal()` vs `require.ElementsMatch()` targets the `kafkacluster_validator_test.go` file where the functions we test return [`field.ErrorList`](https://pkg.go.dev/k8s.io/apimachinery@v0.26.1/pkg/util/validation/field#ErrorList) (which is a `[]*field.Error`).

I'm curious about you guys' opinion on Equal() vs ElementsMatch() because, to rephrase a part of a previous comment, order in that field error list : 
* is deterministic   -->   so using require.Equal() is an option
* but order doesn't matter and it isn't characteristic to the list or its use  -->  so I instinctively chose require.ElementsMatch()
    * also, it makes it a bit easier to write the `expected` results in the test cases since order wouldn't matter

Reply:
Ahhh I see, I looked into the implementation of both methods but still didn't determine which one is better for comparing `field.ErrorList`, so I tried to break the test and see what I get from them:

With `require.Equal()`:

```
=== RUN   TestCheckExternalListenerStartingPort/invalid_config:_3_brokers_with_2_at-the-limit_externalListeners
    kafkacluster_validator_test.go:692: 
        	Error Trace:	/Users/darrelau/go/src/github.com/cisco-app-networking/SDM/koperator/pkg/webhooks/kafkacluster_validator_test.go:692
        	Error:      	Not equal: 
        	            	expected: field.ErrorList{(*field.Error)(0xc0003c33c0), (*field.Error)(0xc0003c3480)}
        	            	actual  : field.ErrorList{(*field.Error)(0xc0003c3800), (*field.Error)(0xc0003c38c0)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -11,3 +11,3 @@
        	            	   BadValue: (int32) 65434,
        	            	-  Detail: (string) (len=169) "invalid external listener starting port number: ExternalListener 'test-external2' would generate invalid port numbers (not between 1 and 65535) for brokers [102] - debug"
        	            	+  Detail: (string) (len=161) "invalid external listener starting port number: ExternalListener 'test-external2' would generate invalid port numbers (not between 1 and 65535) for brokers [102]"
        	            	  })
        	Test:       	TestCheckExternalListenerStartingPort/invalid_config:_3_brokers_with_2_at-the-limit_externalListeners
```

With `require.ElementsMatch()`:
```
    kafkacluster_validator_test.go:692: 
        	Error Trace:	/Users/darrelau/go/src/github.com/cisco-app-networking/SDM/koperator/pkg/webhooks/kafkacluster_validator_test.go:692
        	Error:      	elements differ
        	            	
        	            	extra elements in list A:
        	            	([]interface {}) (len=1) {
        	            	 (*field.Error)({
        	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
        	            	  Field: (string) (len=62) "spec.listenersConfig.externalListeners[1].externalStartingPort",
        	            	  BadValue: (int32) 65434,
        	            	  Detail: (string) (len=169) "invalid external listener starting port number: ExternalListener 'test-external2' would generate invalid port numbers (not between 1 and 65535) for brokers [102] - debug"
        	            	 })
        	            	}
        	            	
        	            	
        	            	extra elements in list B:
        	            	([]interface {}) (len=1) {
        	            	 (*field.Error)({
        	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
        	            	  Field: (string) (len=62) "spec.listenersConfig.externalListeners[1].externalStartingPort",
        	            	  BadValue: (int32) 65434,
        	            	  Detail: (string) (len=161) "invalid external listener starting port number: ExternalListener 'test-external2' would generate invalid port numbers (not between 1 and 65535) for brokers [102]"
        	            	 })
        	            	}
        	            	
        	            	
        	            	listA:
        	            	(field.ErrorList) (len=2) {
        	            	 (*field.Error)({
        	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
        	            	  Field: (string) (len=62) "spec.listenersConfig.externalListeners[0].externalStartingPort",
        	            	  BadValue: (int32) 65535,
        	            	  Detail: (string) (len=164) "invalid external listener starting port number: ExternalListener 'test-external1' would generate invalid port numbers (not between 1 and 65535) for brokers [11 102]"
        	            	 }),
        	            	 (*field.Error)({
        	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
        	            	  Field: (string) (len=62) "spec.listenersConfig.externalListeners[1].externalStartingPort",
        	            	  BadValue: (int32) 65434,
        	            	  Detail: (string) (len=169) "invalid external listener starting port number: ExternalListener 'test-external2' would generate invalid port numbers (not between 1 and 65535) for brokers [102] - debug"
        	            	 })
        	            	}
        	            	
        	            	
        	            	listB:
        	            	(field.ErrorList) (len=2) {
        	            	 (*field.Error)({
        	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
        	            	  Field: (string) (len=62) "spec.listenersConfig.externalListeners[0].externalStartingPort",
        	            	  BadValue: (int32) 65535,
        	            	  Detail: (string) (len=164) "invalid external listener starting port number: ExternalListener 'test-external1' would generate invalid port numbers (not between 1 and 65535) for brokers [11 102]"
        	            	 }),
        	            	 (*field.Error)({
        	            	  Type: (field.ErrorType) (len=17) "FieldValueInvalid",
        	            	  Field: (string) (len=62) "spec.listenersConfig.externalListeners[1].externalStartingPort",
        	            	  BadValue: (int32) 65434,
        	            	  Detail: (string) (len=161) "invalid external listener starting port number: ExternalListener 'test-external2' would generate invalid port numbers (not between 1 and 65535) for brokers [102]"
        	            	 })
        	            	}
        	Test:       	TestCheckExternalListenerStartingPort/invalid_config:_3_brokers_with_2_at-the-limit_externalListeners
```

I think either method works for comparing `field.ErrorList`, but I think `require.Equal()` provides a better test output on failures

Reply:
Sorry for getting here late, being a bit busy.

I think both of you get the right opinion.

Theory: (strict)

If semantically the function is expected to always return the same deterministic order of elements based on the input where there is a relation between the input and the output, I would use `Equal` to signal that the function is expected to comply with that understanding (implicit contract) of returning the elements in predeterminable order - I think of a stable sort for example, but there it is intrinsically part of the semantics of the function, there are less obvious cases like filtering a slice for some elements, in these cases I would suggest using equal.

If the order is not predeterminable or the order of the resulting elements doesn't matter then I would leave it for the test implementor which function is to be used, I wouldn't comment on an `ElementsMatch`, but I personally would favor `Equal` still unless the order of the elements is dynamic in which case I would use `ElementsMatch` - so only use `ElementsMatch` when I absolutely have to, personally. The use-case for this would be when there is a non-stable resorting on elements, for example when there is a weak ordering in a slice - I don't have a concrete example apologies.

Practice: (loose)

In the first case when it is an implicit expectation on ordered elements - input commands being returned in reception order -  I would request changing `ElementsMatch` to `Equal`, in all other cases I would let any of the two fly unless the CI breaks repeatably because of ordering match issues.

Reply:
Also one thing I haven't mentioned.
I really favor the concise output of the `Equal` function in error case so I would prefer that, it is easier to digest by a human afterwards.

Reply:
Thank you both for the explanations!

Commit 4419acde4f82e6fcb7bcf0d23f60631367bcceb5 updates the test file to use `require.Equal()`.